cli/cmd/diagnostics.go

```diff
@@ -1,78 +1,42 @@
-}
-
-// newDiagnosticsOptions initializes diagnostics options setting
-// the max wait time duration as 30 seconds to fetch diagnostics
-//
-// This option may be overridden on the CLI at run-time
-func newDiagnosticsOptions() *diagnosticsOptions {
-	return &diagnosticsOptions{
-		wait: 30 * time.Second,
-	}
-}
-
-// newCmdDashboard creates a new cobra command `diagnostics` which contains commands to fetch control plane container's metrics
+// newCmdDiagnostics creates a new cobra command `diagnostics` which contains commands to fetch Linkerd diagnostics
 func newCmdDiagnostics() *cobra.Command {
-	options := newDiagnosticsOptions()
-
-	cmd := &cobra.Command{
-		Use:   "diagnostics",
-		Short: "Fetch metrics directly from the Linkerd control plane containers",
-		Long: `Fetch metrics directly from Linkerd control plane containers.
-
-  This command initiates port-forward to each control plane process, and
-  queries the /metrics endpoint on them.`,
-		Args: cobra.NoArgs,
-		RunE: func(cmd *cobra.Command, args []string) error {
-			k8sAPI, err := k8s.NewAPI(kubeconfigPath, kubeContext, impersonate, impersonateGroup, 0)
-			if err != nil {
-				return err
-			}
-
-			pods, err := k8sAPI.CoreV1().Pods(controlPlaneNamespace).List(cmd.Context(), metav1.ListOptions{})
-			if err != nil {
-				return err
-			}
-
-			results := getMetrics(k8sAPI, pods.Items, adminHTTPPortName, options.wait, verbose)
-
-			var buf bytes.Buffer
-			for i, result := range results {
-				content := fmt.Sprintf("#\n# POD %s (%d of %d)\n", result.pod, i+1, len(results))
-				if result.err != nil {
-					content += fmt.Sprintf("# ERROR %s\n", result.err)
-				} else {
-					content += fmt.Sprintf("# CONTAINER %s \n#\n", result.container)
-					content += string(result.metrics)
-				}
-				buf.WriteString(content)
-			}
-			fmt.Printf("%s", buf.String())
 
-			return nil
-		},
+	diagnosticsCmd := &cobra.Command{
+		Use:     "diagnostics [flags]",
+		Aliases: []string{"dg"},
+		Args:    cobra.NoArgs,
+		Short:   "Commands used to diagnose Linkerd components",
+		Long: `Commands used to diagnose Linkerd components.
+
+This command provides subcommands to diagnose the functionality of Linkerd.`,
+		Example: `  # Get control-plane component metrics
+  linkerd diagnostics cp-metrics
+
+  # Get metrics from the web deployment in the emojivoto namespace.
+  linkerd diagnostics proxy-metrics -n emojivoto deploy/web
+ 
+  # Get the endpoints for authorities in Linkerd's control-plane itself
+  linkerd diagnostics endpoints linkerd-controller-api.linkerd.svc.cluster.local:8085
+
+  # Install service profiles for the control-plane components.
+  linkerd diagnostics install-sp
+  `,
 	}
 
-	cmd.Flags().DurationVarP(&options.wait, "wait", "w", options.wait, "Time allowed to fetch diagnostics")
+	diagnosticsCmd.AddCommand(newCmdControllerMetrics())
+	diagnosticsCmd.AddCommand(newCmdEndpoints())
+	diagnosticsCmd.AddCommand(newCmdMetrics())
+	diagnosticsCmd.AddCommand(newCmdInstallSP())
```

Comment:
Should we call the `install-sp` something else? As its a bit confusing as this installs service profiles for the control-plane components and not the service profiles crds itself